File: src/utils/migotoLayout.ts
import migotoLayoutManifest from '../shared/migoto-layout.json';

export interface MigotoImporterBehavior {
  defaultUseHook: boolean;
  defaultEnforceRendering: boolean;
  defaultProcessTimeout: number;
  defaultXxmiDllInitDelay: number;
  requiredStartArgs: string[];
  injectionLocked?: boolean;
}

export interface MigotoLockedImporterLegacyDefaults {
  enforceRendering: boolean;
  processTimeout: number;
  xxmiDllInitDelay: number;
}

interface MigotoLayoutManifest {
  defaultImporter: string;
  importerOrder: string[];
  gameImporters: Record<string, string>;
  importerMarkers: string[];
  importerBehaviors: Record<string, MigotoImporterBehavior>;
  defaultImporterBehavior: MigotoImporterBehavior;
  lockedImporterLegacyDefaults: MigotoLockedImporterLegacyDefaults;
}

export interface MigotoPathConfigLike {
  importer?: string | null;
  migoto_path?: string | null;
  importer_folder?: string | null;
  mod_folder?: string | null;
  shader_fixes_folder?: string | null;
  d3dx_ini_path?: string | null;
}

export interface MigotoResolvedPaths {
  importer: string;
  migotoPath: string;
  importerFolder: string;
  modFolder: string;
  shaderFixesFolder: string;
  d3dxIniPath: string;
}

type PathExistsFn = (path: string) => Promise<boolean>;
type JoinPathFn = (base: string, child: string) => string | Promise<string>;

const MIGOTO_LAYOUT = migotoLayoutManifest as MigotoLayoutManifest;

export const trimMigotoPathValue = (value: unknown) =>
  String(value ?? '').trim();

export const normalizeMigotoImporterValue = (value: unknown) =>
  trimMigotoPathValue(value).toUpperCase();

export const splitMigotoStartArgs = (value: unknown) =>
  trimMigotoPathValue(value).split(/\s+/).filter(Boolean);

export const joinMigotoPath = (base: unknown, child: unknown) => {
  const normalizedBase = trimMigotoPathValue(base).replace(/[\\/]+$/g, '');
  const normalizedChild = trimMigotoPathValue(child).replace(/^[/\\]+/g, '');
  if (!normalizedBase) return normalizedChild;
  if (!normalizedChild) return normalizedBase;
  return `${normalizedBase}/${normalizedChild}`;
};

export const getMigotoPathBasename = (path: unknown) => {
  const normalized = trimMigotoPathValue(path).replace(/[\\/]+$/g, '');
  const segments = normalized.split(/[\\/]+/).filter(Boolean);
  return segments[segments.length - 1] || '';
};

export const MIGOTO_DEFAULT_IMPORTER =
  trimMigotoPathValue(MIGOTO_LAYOUT.defaultImporter) || 'WWMI';
export const MIGOTO_IMPORTER_ORDER = [...MIGOTO_LAYOUT.importerOrder];
export const MIGOTO_IMPORTER_MARKERS = [...MIGOTO_LAYOUT.importerMarkers];
export const MIGOTO_DEFAULT_IMPORTER_BEHAVIOR = {
  ...MIGOTO_LAYOUT.defaultImporterBehavior,
};
export const MIGOTO_LOCKED_IMPORTER_LEGACY_DEFAULTS = {
  ...MIGOTO_LAYOUT.lockedImporterLegacyDefaults,
};

export const getRequiredMigotoImporter = (gameName: string) =>
  trimMigotoPathValue(MIGOTO_LAYOUT.gameImporters[gameName] || '');

export const resolveMigotoImporter = (
  gameName: string,
  importerValue: unknown,
) => {
  const requiredImporter = getRequiredMigotoImporter(gameName);
  if (requiredImporter) return requiredImporter;

  const explicitImporter = normalizeMigotoImporterValue(importerValue);
  return explicitImporter || MIGOTO_DEFAULT_IMPORTER;
};

export const getMigotoImporterBehavior = (
  gameName: string,
  importerValue: unknown,
): MigotoImporterBehavior => {
  const importer = resolveMigotoImporter(gameName, importerValue);
  return (
    MIGOTO_LAYOUT.importerBehaviors[normalizeMigotoImporterValue(importer)] ||
    MIGOTO_DEFAULT_IMPORTER_BEHAVIOR
  );
};

const resolveJoinedPath = async (
  base: string,
  child: string,
  joinPath?: JoinPathFn,
) => {
  if (!trimMigotoPathValue(base) || !trimMigotoPathValue(child)) {
    return '';
  }

  if (!joinPath) {
    return joinMigotoPath(base, child);
  }

  return trimMigotoPathValue(await joinPath(base, child));
};

export const pathLooksLikeMigotoImporterFolder = async (
  basePath: string,
  pathExistsAt: PathExistsFn,
  joinPath?: JoinPathFn,
) => {
  const normalizedBase = trimMigotoPathValue(basePath);
  if (!normalizedBase) return false;

  for (const marker of MIGOTO_IMPORTER_MARKERS) {
    try {
      const candidate = await resolveJoinedPath(
        normalizedBase,
        marker,
        joinPath,
      );
      if (candidate && await pathExistsAt(candidate)) {
        return true;
      }
    } catch {
      // ignore invalid candidates and keep probing markers
    }
  }

  return false;
};

export const resolveMigotoImporterFolder = async ({
  basePath,
  importer,
  explicitImporterFolder,
  pathExistsAt,
  joinPath,
}: {
  basePath: string;
  importer: string;
  explicitImporterFolder?: string | null;
  pathExistsAt: PathExistsFn;
  joinPath?: JoinPathFn;
}) => {
  const normalizedExplicit = trimMigotoPathValue(explicitImporterFolder);
  if (normalizedExplicit) {
    return normalizedExplicit;
  }

  const normalizedBase = trimMigotoPathValue(basePath);
  if (!normalizedBase) {
    return '';
  }

  if (
    await pathLooksLikeMigotoImporterFolder(normalizedBase, pathExistsAt, joinPath) ||
    getMigotoPathBasename(normalizedBase).toUpperCase() ===
      normalizeMigotoImporterValue(importer)
  ) {
    return normalizedBase;
  }

  try {
    const nestedImporterFolder = await resolveJoinedPath(
      normalizedBase,
      normalizeMigotoImporterValue(importer),
      joinPath,
    );
    if (nestedImporterFolder && await pathExistsAt(nestedImporterFolder)) {
      return nestedImporterFolder;
    }
  } catch {
    // ignore invalid candidates and fall back to the configured data path
  }

  return normalizedBase;
};

export const buildMigotoResolvedPaths = ({
  gameName,
  config,
  defaultMigotoPath = '',
  detectedImporterFolder = '',
}: {
  gameName: string;
  config: MigotoPathConfigLike;
  defaultMigotoPath?: string;
  detectedImporterFolder?: string;
}): MigotoResolvedPaths => {
  const importer = resolveMigotoImporter(gameName, config.importer);
  const migotoPath =
    trimMigotoPathValue(config.migoto_path) ||
    trimMigotoPathValue(defaultMigotoPath);
  const importerFolder =
    trimMigotoPathValue(config.importer_folder) ||
    trimMigotoPathValue(detectedImporterFolder) ||
    migotoPath;
  const modFolder =
    trimMigotoPathValue(config.mod_folder) ||
    (importerFolder ? joinMigotoPath(importerFolder, 'Mods') : '');
  const shaderFixesFolder =
    trimMigotoPathValue(config.shader_fixes_folder) ||
    (importerFolder ? joinMigotoPath(importerFolder, 'ShaderFixes') : '');
  const d3dxIniPath =
    trimMigotoPathValue(config.d3dx_ini_path) ||
    (importerFolder ? joinMigotoPath(importerFolder, 'd3dx.ini') : '');

  return {
    importer,
    migotoPath,
    importerFolder,
    modFolder,
    shaderFixesFolder,
    d3dxIniPath,
  };
};

export const detectMigotoResolvedPaths = async ({
  gameName,
  config,
  defaultMigotoPath = '',
  pathExistsAt,
  joinPath,
}: {
  gameName: string;
  config: MigotoPathConfigLike;
  defaultMigotoPath?: string;
  pathExistsAt: PathExistsFn;
  joinPath?: JoinPathFn;
}) => {
  const importer = resolveMigotoImporter(gameName, config.importer);
  const migotoPath =
    trimMigotoPathValue(config.migoto_path) ||
    trimMigotoPathValue(defaultMigotoPath);
  const importerFolder = await resolveMigotoImporterFolder({
    basePath: migotoPath,
    importer,
    explicitImporterFolder: config.importer_folder,
    pathExistsAt,
    joinPath,
  });

  return buildMigotoResolvedPaths({
    gameName,
    config,
    defaultMigotoPath,
    detectedImporterFolder: importerFolder,
  });
};

File: src-tauri/src/utils/migoto_layout.rs
use serde::Deserialize;
use serde_json::Value;
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::LazyLock;

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MigotoImporterBehavior {
    pub default_use_hook: bool,
    pub default_enforce_rendering: bool,
    pub default_process_timeout: u32,
    pub default_xxmi_dll_init_delay: u32,
    pub required_start_args: Vec<String>,
    #[serde(default)]
    pub injection_locked: bool,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LockedImporterLegacyDefaults {
    pub enforce_rendering: bool,
    pub process_timeout: u32,
    pub xxmi_dll_init_delay: u32,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
struct MigotoLayoutManifest {
    default_importer: String,
    game_importers: HashMap<String, String>,
    importer_markers: Vec<String>,
    importer_behaviors: HashMap<String, MigotoImporterBehavior>,
    default_importer_behavior: MigotoImporterBehavior,
    locked_importer_legacy_defaults: LockedImporterLegacyDefaults,
}

#[derive(Debug, Clone)]
pub struct MigotoPathState {
    pub importer: String,
    pub migoto_enabled: bool,
    pub migoto_path: PathBuf,
    pub importer_folder: PathBuf,
    pub mod_folder: PathBuf,
    pub shader_fixes_folder: PathBuf,
    pub d3dx_ini_path: PathBuf,
}

static MIGOTO_LAYOUT: LazyLock<MigotoLayoutManifest> = LazyLock::new(|| {
    serde_json::from_str(include_str!("../../../src/shared/migoto-layout.json"))
        .expect("invalid shared migoto layout manifest")
});

pub fn trim_value(value: Option<&str>) -> String {
    value.unwrap_or_default().trim().to_string()
}

fn migoto_section<'a>(config: &'a Value) -> Option<&'a Value> {
    config
        .pointer("/other/migoto")
        .or_else(|| config.get("migoto"))
}

pub fn string_to_path(value: Option<&str>) -> Option<PathBuf> {
    let trimmed = trim_value(value);
    if trimmed.is_empty() {
        None
    } else {
        Some(PathBuf::from(trimmed))
    }
}

pub fn default_importer() -> &'static str {
    let trimmed = MIGOTO_LAYOUT.default_importer.trim();
    if trimmed.is_empty() {
        "WWMI"
    } else {
        trimmed
    }
}

pub fn normalize_importer_name(importer_name: &str) -> String {
    let normalized = importer_name.trim().to_ascii_uppercase();
    if normalized.is_empty() {
        default_importer().to_string()
    } else {
        normalized
    }
}

pub fn required_importer_for_game(game_name: &str) -> Option<String> {
    MIGOTO_LAYOUT
        .game_importers
        .get(game_name)
        .map(|value| value.trim().to_string())
        .filter(|value| !value.is_empty())
}

pub fn resolve_importer(game_name: &str, configured: Option<&str>) -> String {
    if let Some(mapped) = required_importer_for_game(game_name) {
        return normalize_importer_name(&mapped);
    }

    let normalized = configured.unwrap_or_default().trim().to_ascii_uppercase();
    if normalized.is_empty() {
        default_importer().to_string()
    } else {
        normalized
    }
}

pub fn importer_behavior(importer_name: &str) -> MigotoImporterBehavior {
    MIGOTO_LAYOUT
        .importer_behaviors
        .get(&normalize_importer_name(importer_name))
        .cloned()
        .unwrap_or_else(|| MIGOTO_LAYOUT.default_importer_behavior.clone())
}

pub fn locked_importer_legacy_defaults() -> &'static LockedImporterLegacyDefaults {
    &MIGOTO_LAYOUT.locked_importer_legacy_defaults
}

fn path_basename(path: &Path) -> String {
    path.file_name()
        .map(|value| value.to_string_lossy().to_string())
        .unwrap_or_default()
}

pub fn looks_like_migoto_importer_folder(path: &Path) -> bool {
    MIGOTO_LAYOUT
        .importer_markers
        .iter()
        .any(|marker| path.join(marker).exists())
}

pub fn resolve_migoto_importer_folder(
    migoto_data_path: &Path,
    importer_name: &str,
    explicit_importer_folder: Option<&str>,
) -> PathBuf {
    if let Some(explicit) = explicit_importer_folder.filter(|value| !value.trim().is_empty()) {
        return PathBuf::from(explicit);
    }

    let normalized_importer = normalize_importer_name(importer_name);
    if looks_like_migoto_importer_folder(migoto_data_path)
        || path_basename(migoto_data_path).eq_ignore_ascii_case(&normalized_importer)
    {
        return migoto_data_path.to_path_buf();
    }

    let nested_importer_folder = migoto_data_path.join(&normalized_importer);
    if nested_importer_folder.exists() {
        return nested_importer_folder;
    }

    migoto_data_path.to_path_buf()
}

pub fn ensure_required_start_args(start_args: &mut Vec<String>, importer_name: &str) {
    let behavior = importer_behavior(importer_name);
    for required_arg in behavior.required_start_args {
        if start_args
            .iter()
            .any(|arg| arg.eq_ignore_ascii_case(&required_arg))
        {
            continue;
        }
        start_args.push(required_arg);
    }
}

pub fn resolve_migoto_path_state(
    importer_name: &str,
    migoto: Option<&Value>,
    default_migoto_path: PathBuf,
) -> MigotoPathState {
    let importer = normalize_importer_name(importer_name);
    let migoto_path = string_to_path(
        migoto
            .and_then(|value| value.get("migoto_path"))
            .and_then(|value| value.as_str()),
    )
    .unwrap_or(default_migoto_path);

    let importer_folder = resolve_migoto_importer_folder(
        &migoto_path,
        &importer,
        migoto
            .and_then(|value| value.get("importer_folder"))
            .and_then(|value| value.as_str()),
    );

    let mod_folder = string_to_path(
        migoto
            .and_then(|value| value.get("mod_folder"))
            .and_then(|value| value.as_str()),
    )
    .unwrap_or_else(|| importer_folder.join("Mods"));

    let shader_fixes_folder = string_to_path(
        migoto
            .and_then(|value| value.get("shader_fixes_folder"))
            .and_then(|value| value.as_str()),
    )
    .unwrap_or_else(|| importer_folder.join("ShaderFixes"));

    let d3dx_ini_path = string_to_path(
        migoto
            .and_then(|value| value.get("d3dx_ini_path"))
            .and_then(|value| value.as_str()),
    )
    .unwrap_or_else(|| importer_folder.join("d3dx.ini"));

    let migoto_enabled = migoto
        .and_then(|value| value.get("enabled"))
        .and_then(|value| value.as_bool())
        .unwrap_or(false);

    MigotoPathState {
        importer,
        migoto_enabled,
        migoto_path,
        importer_folder,
        mod_folder,
        shader_fixes_folder,
        d3dx_ini_path,
    }
}

pub fn resolve_migoto_path_state_for_game(
    game_name: &str,
    config: &Value,
    default_migoto_path: PathBuf,
) -> MigotoPathState {
    let migoto = migoto_section(config);
    let importer = resolve_importer(
        game_name,
        migoto
            .and_then(|value| value.get("importer"))
            .and_then(|value| value.as_str()),
    );
    resolve_migoto_path_state(&importer, migoto, default_migoto_path)
}

#[cfg(test)]
mod tests {
    use super::{
        ensure_required_start_args, normalize_importer_name, resolve_importer,
        resolve_migoto_importer_folder,
    };
    use std::fs;
    use std::path::PathBuf;
    use std::time::{SystemTime, UNIX_EPOCH};

    fn unique_temp_dir(name: &str) -> PathBuf {
        let unique = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .expect("time")
            .as_nanos();
        std::env::temp_dir().join(format!("ssmt4-migoto-layout-{}-{}", name, unique))
    }

    #[test]
    fn resolves_required_importer_for_known_game() {
        assert_eq!(resolve_importer("ArknightsEndfield", Some("wwmi")), "EFMI");
        assert_eq!(normalize_importer_name(""), "WWMI");
    }

    #[test]
    fn appends_required_start_args_without_duplicates() {
        let mut args = vec!["-dx11".to_string(), "-windowed".to_string()];
        ensure_required_start_args(&mut args, "WWMI");
        assert_eq!(args, vec!["-dx11".to_string(), "-windowed".to_string()]);
    }

    #[test]
    fn uses_nested_importer_folder_when_present() {
        let base = unique_temp_dir("nested");
        let nested = base.join("EFMI");
        fs::create_dir_all(nested.join("Mods")).expect("create nested mods");

        let resolved = resolve_migoto_importer_folder(&base, "EFMI", None);
        assert_eq!(resolved, nested);

        let _ = fs::remove_dir_all(base);
    }
}
